# Notkun gervigreindar fyrir greiningu á þrívíddarmyndum

Nathan Holmes-King

In [28]:
import numpy as np
import pandas as pd
import pywikibot
import sklearn as sk
from sklearn.model_selection import train_test_split
from stl import mesh
import time
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
device = torch.device("cpu")#"mps" if torch.backends.mps.is_available() else "cpu")
torch.set_default_device(device)

## Inngangsorð
Við ætlum að þjálfa gervigreindarlíkan til að greina þrívíddarmyndir. Í notkun eru líkön sem geta það, en þau nota alltaf "bitmap"-myndir. Þetta líkan hér notar "vector"-myndir eins og Envalys.

## Gögn
Þessi gögn eru STL-skrár frá Wikimedia Commons. Það eru fimm flokkar:
- byggingar
- rúmfræði
- geimfarartæki
- styttur

### Sækja gögn

In [29]:
flokkar = ['body parts', 'buildings', 'geometric shapes', 'objects in space', 'sculptures']
skrar = {}
catnum = {}

In [30]:
commons = pywikibot.Site('commons', 'commons')
cn = 0
for a in flokkar:
    print(a)
    cat = pywikibot.Category(commons, 'STL files of ' + a)
    catnum[a] = cn
    cn += 1
    n = 0
    for p in cat.members(member_type=['file']):
        if n % 10 == 0:
            print(n)
        mynd = pywikibot.FilePage(p)
        try:
            tempf = open('/Users/002-nathan/Desktop/Envalys/STLdata/' + a + '_' + p.title()[5:], 'r')
            tempf.close()
        except FileNotFoundError:
            mynd.download(filename='/Users/002-nathan/Desktop/Envalys/STLdata/' + a + '_' + p.title()[5:])
        try:
            skrar[a].append(p.title()[5:])
        except KeyError:
            skrar[a] = [p.title()[5:]]
        n += 1
        if n >= 100:
            break

body parts
0
10
20
30
40
50
60
70
80
buildings
0
10
20
geometric shapes
0
10
20
30
40
objects in space
0
10
20
30
40
50
sculptures
0
10
20
30
40
50


### Setja upp gögn fyrir notkun
Við deilum myndinni í 2.097.152 (128x128x128) þrívíddardíla eða "voxels", teljum punktana í hverjum díl, og notum töluna til að greina myndina.

In [43]:
X_preproc = []
y_preproc = []
for cat in skrar:
    print(cat)
    byrjun = time.time()
    for fi in skrar[cat]:
        # Load data
        gogn = mesh.Mesh.from_file('/Users/002-nathan/Desktop/Envalys/STLdata/' + cat + '_' + fi)
        # Re-scale to be 
        gogn_x = [a[0] for a in gogn.v0]# + [a[0] for a in gogn.v1] + [a[0] for a in gogn.v2]
        gogn_y = [a[1] for a in gogn.v0]# + [a[1] for a in gogn.v1] + [a[1] for a in gogn.v2]
        gogn_z = [a[2] for a in gogn.v0]# + [a[2] for a in gogn.v1] + [a[2] for a in gogn.v2]
        minx = min(gogn_x)
        miny = min(gogn_y)
        minz = min(gogn_z)
        scale = min(128 / (max(gogn_x) + 1e-3 - minx), 
                    128 / (max(gogn_y) + 1e-3 - miny), 
                    128 / (max(gogn_z) + 1e-3 - minz))
        ny_gogn = np.zeros((1, 128, 128, 128), dtype=np.float32)
        for a in gogn.v0:
            x = int((a[0] - minx) * scale)
            y = int((a[1] - miny) * scale)
            z = int((a[2] - minz) * scale)
            ny_gogn[0][x][y][z] += 1
        X_preproc.append(ny_gogn)
        y_preproc.append([catnum[cat]])
    print(time.time() - byrjun)
    print('----')

body parts
45.23072910308838
----
buildings
25.889988899230957
----
geometric shapes
32.926254987716675
----
objects in space
exception (False, 'No lines found, impossible to read')
46.50938415527344
----
sculptures
118.43725895881653
----


In [50]:
X_data = []
y_data = []
for a in X_preproc:
    X_data.append(torch.from_numpy(a).to(device))
for a in y_preproc:
    y_data.append(torch.tensor(a).to(device))

## Líkan

In [51]:
class likan(nn.Module):
    def __init__(self):
        super(likan, self).__init__()
        self.conv1 = nn.Conv3d(1, 16, 5, padding=2)
        self.conv2 = nn.Conv3d(16, 32, 5, padding=2)
        self.conv3 = nn.Conv3d(32, 64, 5, padding=2)
        self.fc1 = nn.Linear(64*16*16*16, 1024)
        self.fc2 = nn.Linear(1024, 128)
        self.fc3 = nn.Linear(128, 5)  # Change second param
    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = F.max_pool3d(x, 2)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool3d(x, 2)
        x = self.conv3(x)
        x = F.relu(x)
        x = F.max_pool3d(x, 2)
        x = x.view(-1, 64*16*16*16)
        x = F.relu(self.fc1(x))
        x = F.dropout(x)
        x = F.relu(self.fc2(x))
        x = F.dropout(x)
        x = self.fc3(x)
        return F.log_softmax(x, dim=1)

In [52]:
num_epochs = 10
learning_rate = 0.01
momentum = 0.5

In [57]:
byrjun = time.time()
acc = []
for i in range(10):
    print('Random state:', i)
    X_data = []
    y_data = []
    for a in X_preproc:
        X_data.append(torch.from_numpy(a).to(device))
    for a in y_preproc:
        y_data.append(torch.tensor(a).to(device))
    X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, random_state=i)
    model = likan()
    optimizer = optim.SGD(model.parameters(), lr=learning_rate, momentum=momentum)
    for n in range(len(X_train)):
        optimizer.zero_grad()
        output = model(X_train[n])
        loss = F.nll_loss(output, y_train[n])
        loss.backward()
        optimizer.step()
    correct = 0
    with torch.no_grad():
        for n in range(len(X_train)):
            output = model(X_train[n])
            y_pred = output.data.max(1, keepdim=True)[1]
            correct += y_pred.eq(y_train[n]).sum().item()
    print('Train accuracy:', correct / len(y_train))
    correct = 0
    with torch.no_grad():
        for n in range(len(X_test)):
            output = model(X_test[n])
            y_pred = output.data.max(1, keepdim=True)[1]
            correct += y_pred.eq(y_test[n]).sum().item()
    print('Test accuracy:', correct / len(y_test))
    acc.append(correct / len(y_test))
    print('Time:', time.time() - byrjun)
    print('----')
print('Mean test accuracy:', np.mean(acc))

Random state: 0
Train accuracy: 0.33507853403141363
Test accuracy: 0.28125
Time: 800.8489389419556
----
Random state: 1
Train accuracy: 0.31413612565445026
Test accuracy: 0.34375
Time: 1575.3194189071655
----
Random state: 2
Train accuracy: 0.3089005235602094
Test accuracy: 0.359375
Time: 2390.0025758743286
----
Random state: 3
Train accuracy: 0.29842931937172773
Test accuracy: 0.390625
Time: 3222.183765888214
----
Random state: 4
Train accuracy: 0.32460732984293195
Test accuracy: 0.3125
Time: 3901.0437047481537
----
Random state: 5
Train accuracy: 0.3193717277486911
Test accuracy: 0.328125
Time: 4568.819064855576
----
Random state: 6
Train accuracy: 0.3298429319371728
Test accuracy: 0.296875
Time: 5248.697355031967
----
Random state: 7
Train accuracy: 0.31413612565445026
Test accuracy: 0.34375
Time: 5953.619192838669
----
Random state: 8
Train accuracy: 0.3193717277486911
Test accuracy: 0.328125
Time: 6699.392108678818
----
Random state: 9
Train accuracy: 0.2879581151832461
Test accur

## Lokaorð

Eftir 1 lotu ("epoch") spáir líkanið rétta flokkinn fyrir á milli 28% og 42% myndanna; meðaltalið er 34%. En ef það væri af handahófi, þá myndi það vera 20%. Líkanið virkar.

Hvernig notum við líkan eins og þetta?
- Til að greina þrívíddarmyndir sem notendur teikna.
- Sem fyrsta skref í stærra líkani sem teiknir sjálft eftir texta sem notendur skrifa.

Vandamál:
- Ekki nóg gögn. Alvörulíkön nota fleiri en 10.000 mynda til að þjálfast.
- Ekki nógir flokkar. Allir flokkar, sérstaklega "líkamshlutar", eru mjög fjölbreyttir og alls ekki eins fyrir tölvuna.
- Við notum aðeins punkta, ekki línur eða flatir, til að greina.

En þetta er það besta fyrir gögnin sem ég var með. Og það tók nóg langan tíma til að þjálfa líkanið; þess vegna þurfum við betri tölvu.